### PR TITLE
PR. Add possibility to retrieve list of projects by pages

### LINKED ISSUE #13: Add possibility to retrieve list of projects by pages
When your list of entities becomes too large (1 million entities for example) it is not good idea to retrieve all of them in one request.
So, we need to improve the method to retrieve list of projects. The method should have two optional parameters "pageNumber", "pageSize".
If the parameters are defined server should send data by correspond filter, otherwise send all of entities.
If one of these parameters is not defined second should not be defined too. It means that if user is defined only one parameter server should return response with 400 code.

### LINKED ISSUE #14: Add possibility to retrieve list of projects sorted by some fields
Client can define a optional URL parameter like this: sort=name:1,id:0 
Where  
“name:1” – first item
“id:0” – second item
So, client can define several fields to sort that spitted by “,”
Each item contains field name and sort direction (0 – ASC, 1 - DESC).

### PR DESCRIPTION
**Pull request by tempo92**
Closes #13

### CODE REVIEW

File: src/main/java/org/tempo/springEdu/contoller/ProjectController.java [hunk 1]
```diff
@@ -30,8 +30,14 @@ String home() {
 
     @GetMapping
     public ResponseEntity<List<ProjectDto>> findAll(
-            @RequestParam(value = "filterProjectName", required = false) String nameFilter) {
-        return ResponseEntity.ok(projectService.findAllDto(nameFilter));
+            @RequestParam(value = "filterProjectName", required = false) String nameFilter
+            ,@RequestParam(value = "pageNumber",
+                required = true, defaultValue = "-1") Integer pageNumber
```
Comment (iv21dk) [inline src/main/java/org/tempo/springEdu/contoller/ProjectController.java:35]: Why `required = true`. In fact the `pageNumber` and `pageSize` still optional

File: src/main/java/org/tempo/springEdu/service/ProjectService.java [hunk 2]
```diff
@@ -48,20 +52,43 @@ public ProjectDto findByIdDto(String id) {
         return entityListToDtoList(repository.findAll());
     }
 
-    public List<ProjectDto> findAllDto(String substring) {
-        if (substring != null && !substring.isEmpty()){
-            return entityListToDtoList(repository.findByAreaRegexIgnoreCase(substring));
+    public List<ProjectDto> findAllDto(
+            String substring, Integer pageNumber, Integer pageSize) {
+
+        if (pageNumber < 0 ^ pageSize == 0) {
```
Comment (iv21dk) [inline src/main/java/org/tempo/springEdu/service/ProjectService.java:58]: Only `null` is cosidered as no value.

File: src/main/java/org/tempo/springEdu/service/ProjectService.java [hunk 3]
```diff
@@ -48,20 +52,43 @@ public ProjectDto findByIdDto(String id) {
...
+            throw new ArgumentException("Parameters 'pageNumber' and 'pageSize' should be set both");
         }
-        else {
-            return findAllDto();
+
+        if (substring != null && !substring.isEmpty()) {
```
Comment (iv21dk) [inline src/main/java/org/tempo/springEdu/service/ProjectService.java:62]: I suggest to use MongoTemplate to build the request. Current implementation looks not so flexible.

**Review by iv21dk — CHANGES_REQUESTED** [review]

File: src/main/java/org/tempo/springEdu/contoller/ProjectController.java (hunk 1 above)
Comment (tempo92) [reply src/main/java/org/tempo/springEdu/contoller/ProjectController.java]: Fixed

File: src/main/java/org/tempo/springEdu/service/ProjectService.java (hunk 2 above)
Comment (tempo92) [reply src/main/java/org/tempo/springEdu/service/ProjectService.java]: Fixed

**Pull request merged**